Checkout: Allow NFC/LNURL-W whenever LNURL is available

dennisreimann opened the pull request:
With what we have in master right now, we display NFC only for top-up invoices. With these changes, we display NFC in all cases, where LNURL is available. Note that this hides LNURL from the list of selectable payment methods, it's only available to use the NFC — and explicitely selectable only for the edge case of top-up invoice + non-unified QR (as before).

Rationale: Now that we got NFC tightly integrated, it doesn't make sense to support the NFC experience only for top-up invoices. With this we bring back LNURL for regular invoices as well, but don't make it selectable and use it only for the NFC functionality.

Closes #4443.

There are many different cases. Basically it's a test matrix permutating:

- Unified/non-unified
- Regular vs. top-up invoice
- LNURL available as payment method

## Unified

### Without LNURL

![unified-without-lnurl](https://user-images.githubusercontent.com/886/220165413-7f31a9db-940c-4367-9bd1-9fd040539f1f.png)

### With LNURL

![unified-with-lnurl](https://user-images.githubusercontent.com/886/220165418-916361d0-8950-4a9f-b64a-be82d2f9632f.png)

### Top-up

![unified-topup](https://user-images.githubusercontent.com/886/220165407-e371d984-ddeb-4ec2-bd86-384de24eb3b6.png)

## Non-unified

### Without LNURL

![non-unified-without-lnurl-bitcoin](https://user-images.githubusercontent.com/886/220267265-4568b44d-1c5e-40d1-b208-0144f92115a1.png) ![non-unified-without-lnurl-lightning png](https://user-images.githubusercontent.com/886/220267277-779e535d-da39-4c98-8ea2-7c968a81846c.png)

### With LNURL

![non-unified-with-lnurl-bitcoin](https://user-images.githubusercontent.com/886/220267365-b8c8b7ec-d883-4f4a-a7af-11285710d846.png) ![non-unified-with-lnurl-lightning](https://user-images.githubusercontent.com/886/220267349-3f24ba25-df9d-4ff1-822d-273a0bf8df15.png)

### Top-up

![non-unified-topup-bitcoin](https://user-images.githubusercontent.com/886/220268551-b819a2db-badb-46d6-9243-d952ec0f8d59.png) ![non-unified-topup-lnurl](https://user-images.githubusercontent.com/886/220268584-c572e088-1c8f-41df-9498-f761e542f5c0.png)

## Lightning with LNURL

![ln-with-lnurl](https://user-images.githubusercontent.com/886/220165547-113c0f6c-81c7-46aa-a8eb-3c1c181f759d.png)

## Just Lightning without LNURL

![lightning-without-lnurl](https://user-images.githubusercontent.com/886/220391033-91d192c7-fbd2-4e68-a34e-4542bec0f97d.png)

dennisreimann commented on BTCPayServer/Controllers/UIInvoiceController.UI.cs:
```diff
@@ -703,9 +705,13 @@ public async Task<IActionResult> CheckoutNoScript(string? invoiceId, string? id
+                        return paymentMethodId.CryptoCode == pmId.CryptoCode && pmId != lnurlId;
+                    });
                 if (paymentMethodTemp == null)
-                    paymentMethodTemp = invoice.GetPaymentMethods().FirstOrDefault();
```
This and the part above prevents LNURL from showing up as initial payment method. We don't want this, as LNURL also isn't explicitely selectable.

dennisreimann commented on BTCPayServer/Controllers/UIInvoiceController.UI.cs:
```diff
@@ -768,6 +774,7 @@ public async Task<IActionResult> CheckoutNoScript(string? invoiceId, string? id
                 BrandColor = storeBlob.BrandColor,
                 CheckoutType = invoice.CheckoutType ?? storeBlob.CheckoutType,
                 HtmlTitle = storeBlob.HtmlTitle ?? "BTCPay Invoice",
+                OnChainWithLnInvoiceFallback = storeBlob.OnChainWithLnInvoiceFallback,
```
We now need this to discern on the UI-level, whether or not to have LNURL available as payment method — which is the case only for top-up + non-unified.

dennisreimann commented on BTCPayServer/Controllers/UIInvoiceController.UI.cs:
```diff
@@ -803,9 +810,6 @@ public async Task<IActionResult> CheckoutNoScript(string? invoiceId, string? id
                 IsMultiCurrency = invoice.GetPayments(false).Select(p => p.GetPaymentMethodId()).Concat(new[] { paymentMethod.GetId() }).Distinct().Count() > 1,
                 StoreId = store.Id,
                 AvailableCryptos = invoice.GetPaymentMethods()
-                                          .Where(i => i.Network != null && storeBlob.CheckoutType == CheckoutType.V1 ||
-                                              // Exclude LNURL for Checkout v2 + non-top up invoices
-                                              i.GetId() != lnurlId || invoice.IsUnsetTopUp())
```
Instead of completely removing LNURL for non-topup invoices, this leaves LNURL in, but hides it from the payment method switches.

dennisreimann commented on BTCPayServer/Views/Shared/NFC/CheckoutEnd.cshtml:
```diff
@@ -26,9 +25,7 @@ Vue.component("lnurl-withdraw-checkout", {
-            return (
-                this.model.paymentMethodId === 'BTC_LNURLPAY' || (
-                    this.model.paymentMethodId === 'BTC' && this.model.invoiceBitcoinUrl.match(/lightning=lnurl/i)));
+              return !!this.model.availableCryptos.find(pm => pm.paymentMethodId === 'BTC_LNURLPAY');
```
Allows NFC whenever LNURL is available.

dennisreimann commented:
> I am for completely removing it where it makes no sense, graying it out, may just add confusion. Oh, could be that you're actually already doing that removing lnurl button when estimated that it's not functional or just per payment method?

@pavlenex Yes, it already isn't available in Checkout v2 anymore — we completely removed it for non-top-up invoices. And this PR reintroduced it kinda in "hidden" mode, so that it is available for NFC beyond just top-up invoices.

In that sense we kinda went to far by removing it completely, because without the changes here, LNURL (and hence NFC) wouldn't be available for regular invoices. I think the above screenshots explain it better — LNURL isn;t shown as a payment method other than for the top-up + non-BIP21 case.

pavlenex commented:
Got it, I have to test this and sync up, just to make sure I understood correctly, if the invoice is 10$, and lightning is available as a payment method, all other criterias (browser, etc) are met, the user would be able to checkout by tapping let's say boltcard? I just want to be completely sure I understand 

> we completely removed it for non-top-up invoices

NicolasDorier commented:
Code ACK for me, however, I am a bit overwhelmed by all of this, so I'll let  @pavlenex  judge :D

dennisreimann commented:
@NicolasDorier yeah, we'll need proper testing. I just turned this into a draft again, because I noticed that it still isn't fully correct and breaks again what I fixed in #4660. 

Will also add checks for the button in the Selenium tests.

dennisreimann commented:
> we'll need proper testing. I just turned this into a draft again, because I noticed that it still isn't fully correct and breaks again what I fixed in https://github.com/btcpayserver/btcpayserver/pull/4660.

Fixed that, added test cases and updated the screenshots above.

dennisreimann commented:
> just to make sure I understood correctly, if the invoice is 10$, and lightning is available as a payment method, all other criterias (browser, etc) are met, the user would be able to checkout by tapping let's say boltcard?

@pavlenex With the changes in this PR - yes, and that's the purpose of this pull request. Before it wouldn't show. 

Screenshots for your test case:

## This PR

![pav-pr](https://user-images.githubusercontent.com/886/220270173-998bbacf-7e60-4e41-946d-91de3f0c2320.png)

## On current master

![pav-master](https://user-images.githubusercontent.com/886/220270166-5491b355-5b3c-40f4-922c-5ceabdfc58a5.png)

ndeet commented:
Nice work! 

When I tested NFC with checkout v1 it was a bit annoying to always scroll down to have the NFC button show up. Maybe a bit OT and for a follow up:
I think it would be useful to have an option to show the NFC button more at the top. If someone runs a pos terminal and mostly has face2face customers, they either scan the QR or pay with NFC. The "pay in wallet" and "LN invoice" button/text is not really useful to them. So maybe a "pos mode" for checkout?

NicolasDorier commented:
@ndeet I think we should eventually have a Checkout "Retail mode" automatically stripping stuff. Like the `Pay in wallet`.

Maybe not for this PR, but I think checkout pages could get more optimized for retails.

dennisreimann commented:
@NicolasDorier Yes, I still have it on my list for v1.8.x — didn't get to it in time for this release.

Kukks reviewed (changes requested):
the NFC option is available regardless of LNURL. If there is a bolt11 or an lnurl on the screen, it should be shown.

dennisreimann commented:
@Kukks thanks for the pointer, I've fixed this in [3474fd5](https://github.com/btcpayserver/btcpayserver/pull/4671/commits/3474fd5cd80b8f63097a90d099af38ad9ac12f02) and also added a test for that case.

Kukks commented on BTCPayServer/Views/UIInvoice/CheckoutV2.cshtml:
```diff
@@ -11,7 +11,10 @@
     var hasPaymentPlugins = UiExtensions.Any(extension => extension.Location == "checkout-payment-method");
-    var paymentMethodCount = Model.AvailableCryptos.Count;
+    // Show LNURL as selectable payment method only for top up invoices + non-BIP21 case
+    var displayedPaymentMethods = Model.IsUnsetTopUp && !Model.OnChainWithLnInvoiceFallback
```
We should really make this glue more flexible so thta we dint hardcode  these things in a view
